# Lab 2: Linear regression
---

Import libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from   scipy import stats

Load the data set in the companion **demodataset.csv**

In [2]:
data = np.loadtxt("demodataset.csv", delimiter=",")

np.set_printoptions(precision=5)
np.set_printoptions(suppress=True)
data

array([[  -0.878,  -32.208],
       [   1.36 ,   49.178],
       [   1.64 ,   63.384],
       [   0.542,   50.955],
       [   0.825,   76.403],
       [  -0.596,  -90.715],
       [  -0.056,   30.239],
       [  -0.132,  -29.705],
       [  -2.435, -139.298],
       [  -0.404,  -10.483],
       [   0.382,   34.85 ],
       [   1.367,  103.15 ],
       [  -0.748,  -25.526],
       [  -0.382,    7.506],
       [   0.842,   52.244],
       [  -0.077,    7.081],
       [   0.433,   76.325],
       [  -0.217,  -52.17 ],
       [   1.468,   71.62 ],
       [   0.113,   -0.239],
       [   1.731,   17.363],
       [  -0.336,  -19.43 ],
       [  -1.099,  -39.712],
       [  -0.842,  -71.2  ],
       [   0.566,   25.936],
       [  -1.058,  -65.185],
       [   0.695,    7.989],
       [   0.539,   25.539],
       [   0.502,   55.364],
       [  -1.439,  -83.49 ],
       [  -0.189,   -2.155],
       [  -0.074,    4.711],
       [  -1.793,  -19.926],
       [   2.231,  105.28 ],
       [   0.0

---
Extract the first column in the data set to vector **x** and the second column to vector **y**.   

Plot a figure showing the dataset as: { $(x^{(i)},(x^{(i)}) | i=1,…, m $}

In [3]:
x=data[:,0] # x is the first  column in the data set
y=data[:,1] # y is the second column in the data set


---
1. a) Compute $θ_0^*$ and $θ_1^*$ for the line of best fit. \
\
i) using scipy.stats.linregress():

In [4]:
Θ1, Θ0, r, p, std_err = stats.linregress(x, y)  # theta1 = slope, theta0 = intercept y axis
print('Θ0 (intercept) = {:.3f}'.format(Θ0))
print('Θ1 (slope)     = {:.3f}'.format(Θ1))

Θ0 (intercept) = 0.802
Θ1 (slope)     = 56.758


1 a) \
\
ii)	implementing the linear regression model below using numpy functions

1 a)\
\
iii)	check that i) and ii) produce the same results

---
1. b) Superimpose the line of best fit to the data

---
1. c)	Predict y for x = 2.231. What is the corresponding residual? Superimpose on the graphic obtained in b)

---
2. a) \
\
i) Compute $J(θ_0^*, θ_1^*)$ for the dataset.

---
2. a)\
\
i) Express the cost function: \
\
 $\qquad J(θ_0, \dots , θ_n) = \frac{1}{2m}\sum_{i=1}^{m} (h_θ(\mathbf x^{(i)}) - y^{(i)})^2$ \
\
in vectorial notation and compute it for the given dataset

- Considering a dataset with $m$ data points (or samples), $n$ features and 1 target or label. \
$\mathbf X$ matrix has $m$ rows, one for each data point and $1 + n$ columns, the left most for $bias$ and the rest for the features. \
$\mathbf y$ vector has $m$ rows, one for each target: \
\
$
\qquad 
\begin{split}
\mathbf{X}_{m\times (1+n)} &=
\begin{bmatrix}
    1 & x_1^{(1)}   & x_2^{(1)} & \dots & x_n^{(1)}\\
    1 & x_1^{(2)}   & x_2^{(2)} & \dots & x_n^{(2)}\\
    & & \vdots & &\\
    1 & x_1^{(m)} & x_2^{(m)} & \dots & x_n^{(m)}\\
\end{bmatrix}
\quad\quad
\mathbf{y} &=
\begin{bmatrix}
    y^{(1)}\\
    y^{(2)}\\
    \vdots \\
    y^{(m)}\\
\end{bmatrix}
\end{split} 
$

$\qquad$ and vector $\boldsymbol\theta$ for $n$ coefficients:

$
\qquad
\boldsymbol{\theta} = \begin{bmatrix}
θ_0 \\ θ_1 \\ \vdots \\ \theta_n
\end{bmatrix}
$

- The hypothesis function, for each $h^{(i)}$, where
$
\mathbf x^{(i)} = \begin{bmatrix}
1 & x_1^{(i)} & \dots & x_n^{(i)}
\end{bmatrix}
$
, given by: \
\
$$h_\theta(\mathbf x^{(i)}) = θ_0.1 + θ_1x_1^{(i)} + θ_2x_2^{(i)} + \dots + θ_2x_n^{(i)} $$

-  can be rewritten in vector form as the dot product of:

$$
\begin{split}
{h_\theta}(\mathbf x^{(i)}) &= 
\begin{bmatrix}
1 & x_1^{(i)} & \dots & x_n^{(i)}
\end{bmatrix}
\begin{bmatrix}
θ_0 \\ θ_1 \\ \vdots \\ \theta_n
\end{bmatrix} \\
{h_\theta}(\mathbf x^{(i)}) &= \mathbf x^{(i)}\boldsymbol{\theta}
\end{split}
$$

- and for $m$ data points, as:
$$
\begin{split}
h_θ(\mathbf{X}) &= \mathbf{X}\boldsymbol{\theta}
\end{split}
$$

- this will compute vector $\mathbf{h}$ with $m$ elements:
  
$$\mathbf{h} = \mathbf{X}\boldsymbol{\theta}$$

- Now, expressing the cost (or loss) function: 

$$J(θ_0, \dots , θ_n) = \frac{1}{2m}\sum_{i=1}^{m} (h_θ(\mathbf x^{(i)}) - y^{(i)})^2$$

- in vector notation:

$$
\begin{split}
J(θ_0, \dots , θ_n) &= \frac{1}{2m} (\mathbf{h} - \mathbf{y})^2 \\
J(θ_0, \dots , θ_n) &= \frac{1}{2m} (\mathbf{h} - \mathbf{y})^T(\mathbf{h} - \mathbf{y})
\end{split}
$$

- computing the residuals vector with:

$\qquad \qquad \mathbf{r} = \mathbf{h} - \mathbf{y}$

$$
  J(θ) = J(θ_0, \dots, θ_n) = \color{red}{\text{Enter final answer here}}
$$

---
compute $J(θ_0^*, θ_1^*)$ in vector notation, for the given dataset. 

---
2. a)\
\
ii)	Is there any advantage to computing in vector notation with NumPy?

$\color{red}{\text{Enter answer here:}}$

\
\
\
.


---
2. a)\
\
iii)	Show that point (average of X, average of Y) belongs to the line of best fit.

$\color{red}{\text{Write the math proof here:}}$

\
\
\
.


---
2. b) For $θ_0^*$ and $θ_1^*$, plot the residuals vs the independent variable x. Print the mean and variance of the residuals. Briefly comment on what you observe.

Residuals are the difference between the observed (y) and predicted responses (h). 

If the model is appropriate for the data, this should be reflected in the residuals. In particular, the graph of
residuals vs. independent variable should be centered in zero without any tendency:

The closer the residuals are from zero the smaller the cost J.

For the normal error regression model, it is assumed that the error term is normally distributed with null mean and
constant variance The mean and variance of the residuals are: 

---
2. c) For $θ_0^*$ and $θ_1^*$, analytically prove that the sum of residuals is zero. Print the sum of the residuals.



- ie. show that: $\qquad\qquad\qquad\qquad \sum_{i=1}^{m} \left( h_θ(x^{(i)}) - y^{(i)} \right) = 0 \qquad\qquad\qquad for: \qquad θ_0^*$,&emsp;$θ_1^*$

- since:

  $\qquad\qquad h_θ(x) = θ_0^* + θ_1^*x$

$$
\begin{split}
    \sum_{i=1}^{m} \left( h_θ(x^{(i)}) - y^{(i)}  \right) = \sum_{i=1}^{m} \left[ (θ_0^* + θ_1^*x)-y^{(i)} \right]
\end{split}
$$

- $θ_0^*$ is constant, so:
$$
\begin{split}
    &= mθ_0^*+ \sum_{i=1}^{m}(θ_1^*x^{(i)}-y^{(i)}) \\
    &= mθ_0^* + θ_1^* \sum_{i=1}^{m} x^{(i)} - \sum_{i=1}^{m} y^{(i)} 
\end{split}
$$

$\color{red}{\text{To complete ...}}$


---
The Sum of the residuals is:

---
2. d) Prove that the derivative of the cost function in order to $\theta_j$ is:

$$
\begin{split}
& \frac{\partial}{\partial \theta_j} J(θ_0, \dots, θ_n) = 
 \frac{1}{m} \sum_{i=1}^{m} (h_\theta(x_j^{(i)})-y^{(i)})   x_j^{(i)}\quad\text,\quad j=0,1 \dots n
\end{split}
$$

The derivative of the cost function in order to $\theta_j$: 

$$
\begin{split}
& \frac{\partial}{\partial \theta_j} \quad J(θ_0, \dots, θ_n) \\
& \frac{\partial}{\partial \theta_j} \quad \frac{1}{2m}\sum_{i=1}^{m} (h_θ(x_j^{(i)}) - y^{(i)})^2 \\
\end{split}
$$

using the derivative of the product rule: 

$\qquad (f.g)' = f'.g + f.g'$

and assuming: 

$
\begin{split}
& \quad f = \frac{1}{2m} \\
& \quad g = \sum_{i=1}^{m}(h_θ(x_j^{(i)}) - y^{(i)})^2
\end{split}
$

since it is a partial derivative in order to $\theta_j, f$ is constant, $f'$ is zero:

we have: $\qquad 0.g + f.g'$

now, using the derivative of the sum rule:

$
\begin{split}
 \qquad (f_0+f_1+\dots+f_n)' &= f_0'+f_1'+\dots+f_n' \\
 \quad (\sum_{i=1}^{m} f)' &= \sum_{i=1}^{m} f'
\end{split}
$

we got:

$$
\frac{1}{2m} \quad \sum_{i=1}^{m} \quad \frac{\partial}{\partial \theta_j} (h_θ(x_j^{(i)}) - y^{(i)})^2 
$$

using the derivative of the power rule:

$\qquad (g^n)' = n.g^{n-1}.g'$

we got:

$$
\frac{1}{\cancel 2 m} \quad \sum_{i=1}^{m} \quad \cancel 2.(h_θ(x_j^{(i)}) - y^{(i)}) \quad \frac{\partial}{\partial \theta_j} (h_θ(x_j^{(i)}) - y^{(i)}) 
$$

\
$\color{red}{\text{To complete ...}}$



List of derivative rules:

https://www.mathsisfun.com/calculus/derivatives-rules.html

---
3. a) Express in vector notation the following gradient descent updating expressions: </p>
$$\theta_j = \theta_j - \alpha \frac{1}{m} 
\sum_{i=1}^{m} (h_\theta(x_j^{(i)})-y^{(i)})   x_j^{(i)}\quad\text,\quad j=0,1 \dots n$$

using $\mathbf{h}$, $\mathbf{y}$, $\boldsymbol{\theta}$ and $\mathbf{X}$ as defined in 2. a) ii):

where it is also shown that:

$$\sum_{i=1}^{m} (h_\theta(x_j^{(i)})-y^{(i)}) = \mathbf{h}-\mathbf{y}$$

$\color{red}{\text{To complete ...}}$

---
3. b) Apply gradient descent and linear regression to the dataset. Find $\theta_0*, \theta_1*$ and $J(\theta_0^*,\theta_1^*)$.

---
3. c) What is the number of iterations and the value of the learning rate $\alpha$, that approximate the $\theta$ vector and the cost J obtained in 3. b) to those found in 1. a) and 2. a)?

$\color{red}{\text{Answer here ...}}$

---
3. d)	Plot $J(\theta_0,\theta_1)$ as a function of the number of iterations. Briefly comment on what you observe.